### PR TITLE
On Demand Druid Exhaust for Password Protected CSV Story

### PR DESCRIPTION
**Pull request by Shakthieshwari**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-24105" title="SB-24105" target="_blank">StorySB-24105</a>  Password Protected CSV
  </td></table>
  <br />
 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+import java.util.function.Supplier
+
+
+case class RequestBody(`type`: String,`params`: Map[String,AnyRef])
+case class DruidQuery(queryType: String, dataSource: String, intervals: String, granularity: Option[String], aggregations: Option[List[Aggregation]],
+                      dimensions: Option[List[DruidDimension]], filters: Option[List[mutable.Map[String,AnyRef]]], having: Option[DruidHavingFilter],
+                      postAggregation: Option[List[PostAggregation]], columns: Option[List[String]], sqlDimensions: Option[List[DruidSQLDimension]],
+                      threshold: Option[Long], metric: Option[String], descending: Option[String], intervalSlider: Int)
+case class Metrics(metric : String, label : String, druidQuery : DruidQuery)
+case class ReportConfig(id: String, queryType: String, dateRange: QueryDateRange, metrics: List[Metrics], labels: LinkedHashMap[String, String],
```
Comment (sowmya-dixit) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: ReportConfig, Metrics and DruidQuery case classes can be directly used by importing from Framework/CoreDataProducts and no need to define once again here.

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+  def validateEncryptionRequest(request: JobRequest): Boolean = {
+    if (request.encryption_key.nonEmpty) true else false;
+  }
+
+  def validateRequest(request: JobRequest): Boolean = {
+    if (request.request_data != null) true else false
+  }
+
+  def getRequests(jobId :String)(implicit spark: SparkSession, fc: FrameworkContext): Array[JobRequest] = {
```
Comment (sowmya-dixit) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: getRequests method already available in OnDemandExhaustJob - why is this implemented again here?

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    request.status = "PROCESSING";
+    updateStatus(request);
+  }
+
+  def getStringProperty(config: Map[String, AnyRef], key: String, defaultValue: String) : String = {
+    config.getOrElse(key, defaultValue).asInstanceOf[String]
+  }
+  def druidAlgorithm(reportConfig: ReportConfig)(implicit spark: SparkSession, sqlContext: SQLContext, fc: FrameworkContext, sc:SparkContext, config: JobConfig): RDD[DruidOutput]  ={
+    val strConfig = JSONUtils.serialize(reportConfig)
+    val reportConfigs = JSONUtils.deserialize[ReportConfig](strConfig)
```
Comment (sowmya-dixit) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Why is reportConfig serialised and again deserialised back to same type **ReportConfig** here? reportConfig can be directly used instead.

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+      else {
+        request.status = "SUCCESS";
+        request.download_urls = Option(response.file);
+        request.execution_time = Option(result._1);
+        request.dt_job_completed = Option(System.currentTimeMillis)
+      }
+    }
+    request
+  }
+  def saveRequestAsync(request: JobRequest,jobId:String)(implicit conf: Configuration, fc: FrameworkContext): CompletableFuture[JobRequest] = {
```
Comment (sowmya-dixit) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Same with saveRequestAsync - this method already available in OnDemandExhaustJob - why is this implemented again here?

File: data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+
+    OnDemandDruidExhaustJob.execute()
+    val postgresQuery = EmbeddedPostgresql.executeQuery("SELECT * FROM job_request WHERE job_id='ml-task-detail-exhaust'")
+    while(postgresQuery.next()) {
+      postgresQuery.getString("status") should be ("SUCCESS")
+      postgresQuery.getString("requested_by") should be ("36b84ff5-2212-4219-bfed-24886969d890")
+      postgresQuery.getString("requested_channel") should be ("ORG_001")
+      postgresQuery.getString("err_message") should be ("")
+      postgresQuery.getString("iteration") should be ("0")
+      postgresQuery.getString("encryption_key") should be ("test@123")
```
Comment (sowmya-dixit) [inline data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala:116]: Also add assertions on download_url field

File: data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+        |"format": "csv","quoteColumns": ["Role","Declared State","District","Block","School Name","Organisation Name","Program Name","Project Title",
+        |"Project Objective","Category","Tasks","Sub-Tasks","Remarks"]},"output":[{"to":"file","params":{"file":"src/test/resources/ml-druid-data-detail"}}],
+        |"parallelization":8,"appName":"ML Druid Data Model"}""".stripMargin
+    val jobConfig = JSONUtils.deserialize[JobConfig](strConfig)
+    implicit val config = jobConfig
+    implicit val conf = spark.sparkContext.hadoopConfiguration
+    OnDemandDruidExhaustJob.execute()
+    val postgresQuery = EmbeddedPostgresql.executeQuery("SELECT * FROM job_request WHERE job_id='ml-task-detail-exhaust'")
+    while (postgresQuery.next()) {
+      postgresQuery.getString("status") should be ("FAILED")
```
Comment (sowmya-dixit) [inline data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala:238]: Add assertions on err_message. Will error gives proper message indicating No Interval?

**Review by sowmya-dixit — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 1 above)
Comment (Ashwiniev95) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: ReportConfig changes is done. other part we are working.

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 2 above)
Comment (Ashwiniev95) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: We Fixed this.

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 1 above)
Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Metrics and DruidQuery case classes is also done

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 3 above)
Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Fixed this issue

File: data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala (hunk 5 above)
Comment (Ashwiniev95) [reply data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala:116]: Added the Assertions

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 4 above)
Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: As discussed with you , we renamed the method with proper naming convention

File: data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala (hunk 6 above)
Comment (Ashwiniev95) [reply data-products/src/test/scala/org/sunbird/analytics/exhaust/TestOnDemandDruidExhaustJob.scala:238]: Added Assertion. err_message will be "Invalid request"

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+  def druidPostProcess(data: RDD[DruidOutput],request_id: String,reportConfig: ReportConfig, storageConfig:StorageConfig)(implicit spark: SparkSession, sqlContext: SQLContext, fc: FrameworkContext, sc:SparkContext, config: JobConfig): OnDemandDruidResponse = {
+    val labelsLookup = reportConfig.labels
+    implicit val sqlContext = new SQLContext(sc)
+    val dimFields = reportConfig.metrics.flatMap { m =>
+      if (m.druidQuery.dimensions.nonEmpty) m.druidQuery.dimensions.get.map(f => f.aliasName.getOrElse(f.fieldName))
+      else if(m.druidQuery.sqlDimensions.nonEmpty) m.druidQuery.sqlDimensions.get.map(f => f.fieldName)
+      else List()
+    }
+    val dataCount = sc.longAccumulator("DruidReportCount")
+    val filesMutable = scala.collection.mutable.MutableList[String]()
```
Comment (sowmya-dixit) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: As updated earlier, this filesMutable variable can be avoided.

Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Fixed this logic

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+  def markRequestAsProcessing(request: JobRequest):Boolean = {
+    request.status = "PROCESSING";
+    updateStatus(request);
+  }
+
+  def getStringProperty(config: Map[String, AnyRef], key: String, defaultValue: String) : String = {
+    config.getOrElse(key, defaultValue).asInstanceOf[String]
+  }
+  def druidAlgorithm(reportConfig: ReportConfig)(implicit spark: SparkSession, sqlContext: SQLContext, fc: FrameworkContext, sc:SparkContext, config: JobConfig): RDD[DruidOutput]  ={
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: One implicit sparkSession is sufficient. sqlContext and sparkContext are redundant

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+      }
+    }
+    val finalResult = metrics.fold(sc.emptyRDD)(_ union _)
+    finalResult.map { f =>
+      DruidOutput(f._2)
+    }
+  }
+
+  def druidPostProcess(data: RDD[DruidOutput],request_id: String,reportConfig: ReportConfig, storageConfig:StorageConfig)(implicit spark: SparkSession, sqlContext: SQLContext, fc: FrameworkContext, sc:SparkContext, config: JobConfig): OnDemandDruidResponse = {
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Same as above. only have spark session as implicit variable

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    val labelsLookup = reportConfig.labels
+    implicit val sqlContext = new SQLContext(sc)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Why is this required? why is this declared if it is available as implicit variable?

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        val res = CommonUtil.time(updateRequest(request))
+        JobLogger.log("Request is zipped", Some(Map("requestId" -> request.request_id, "timeTakenForZip" -> res._1)), INFO)
+        request
+      }
+    })
+  }
+
+  def execute()(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig, sqlContext:SQLContext, sc:SparkContext, conf: Configuration): Metrics = {
+    val reportConfig = JSONUtils.deserialize[ReportConfig](JSONUtils.serialize(config.modelParams.get("reportConfig")))
+    val jobId : String = reportConfig.id
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: The report configurations should be a json map right? We should fetch the report config with the job id.

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+    val finalReportConfig = JSONUtils.deserialize[ReportConfig](finalConfig)
+    val druidData : RDD[DruidOutput] = druidAlgorithm(finalReportConfig)
+    val result = CommonUtil.time(druidPostProcess(druidData,request.request_id,JSONUtils.deserialize[ReportConfig](JSONUtils.serialize(finalReportConfig)),storageConfig))
+    val response = result._2;
+    val failedOnDemandDruidRes = response.status.equals("FAILED")
+    if (failedOnDemandDruidRes) {
+      markRequestAsFailed(request, response.statusMsg)
+    } else {
+      if(validateEncryptionRequest(request) == true){
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: No need of an equals to comparison

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 8 above)
Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: sqlContext is removed , But implicit sparkContext is required to call DruidDataFetcher.getDruidData()

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 9 above)
Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: implicit sc:SparkContext is required => as we use it ,to get the dataCount (val dataCount = sc.longAccumulator("DruidReportCount")) & getReportDF() requires it
       implicit sqlContext: SQLContext is required => as getReportDF() requires

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 10 above)
Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Fixed it

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 12 above)
Comment (Shakthieshwari) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Fixed it

File: data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala (hunk 11 above)
Comment (Ashwiniev95) [reply data-products/src/main/scala/org/sunbird/analytics/exhaust/OnDemandDruidExhaustJob.scala]: Fixed it

**Review by SanthoshVasabhaktula — APPROVED** [review]
